backend/gce.go

```diff
@@ -539,62 +468,131 @@ func (p *gceProvider) Start(ctx gocontext.Context, startAttributes *StartAttribu
+	errChan := state.Get("err_chan").(chan error)
 
-	var (
-		image *compute.Image
-		err   error
-	)
+	scriptBuf := bytes.Buffer{}
+	err := gceStartupScript.Execute(&scriptBuf, p.ic)
+	if err != nil {
+		errChan <- err
+		return multistep.ActionHalt
+	}
 
-	candidateLangs := []string{}
+	state.Put("script", scriptBuf.String())
+	return multistep.ActionContinue
+}
 
-	mappedLang := fmt.Sprintf("LANGUAGE_MAP_%s", strings.ToUpper(startAttributes.Language))
-	if p.cfg.IsSet(mappedLang) {
-		logger.WithFields(logrus.Fields{
-			"original": startAttributes.Language,
-			"mapped":   p.cfg.Get(mappedLang),
-		}).Debug("using mapped language to candidates")
-		candidateLangs = append(candidateLangs, p.cfg.Get(mappedLang))
-	} else {
-		logger.WithFields(logrus.Fields{
-			"original": startAttributes.Language,
-		}).Debug("adding original language to candidates")
-		candidateLangs = append(candidateLangs, startAttributes.Language)
+func (p *gceProvider) stepInsertInstance(state multistep.StateBag) multistep.StepAction {
+	errChan := state.Get("err_chan").(chan error)
+	image := state.Get("image").(*compute.Image)
+	script := state.Get("script").(string)
+	startAttributes := state.Get("start_attributes").(*StartAttributes)
+	ctx := state.Get("ctx").(gocontext.Context)
+
+	inst := p.buildInstance(startAttributes, image.SelfLink, script)
+
+	context.LoggerFromContext(ctx).WithFields(logrus.Fields{
+		"instance": inst,
+	}).Debug("inserting instance")
+
+	state.Put("boot_start", time.Now().UTC())
+
+	op, err := p.client.Instances.Insert(p.projectID, p.ic.Zone.Name, inst).Do()
+	if err != nil {
+		errChan <- err
+		return multistep.ActionHalt
 	}
-	candidateLangs = append(candidateLangs, p.defaultLanguage)
 
-	for _, language := range candidateLangs {
-		logger.WithFields(logrus.Fields{
-			"original":  startAttributes.Language,
-			"candidate": language,
-		}).Debug("searching for image matching language")
+	state.Put("instance", inst)
+	state.Put("instance_insert_op", op)
+	return multistep.ActionContinue
+}
+
+func (p *gceProvider) stepWaitForInstanceIP(state multistep.StateBag) multistep.StepAction {
+	errChan := state.Get("err_chan").(chan error)
+	instChan := state.Get("inst_chan").(chan Instance)
+	inst := state.Get("inst").(*compute.Instance)
+	image := state.Get("image").(*compute.Image)
+	startBooting := state.Get("boot_start").(time.Time)
+	op := state.Get("instance_insert_op").(*compute.Operation)
+	ctx := state.Get("ctx").(gocontext.Context)
+	logger := context.LoggerFromContext(ctx)
+
+	logger.WithFields(logrus.Fields{
+		"duration": p.bootPrePollSleep,
+	}).Debug("sleeping before first checking instance insert operation")
+
+	time.Sleep(p.bootPrePollSleep)
+
+	zoneOpCall := p.client.ZoneOperations.Get(p.projectID, p.ic.Zone.Name, op.Name)
+
+	for {
+		newOp, err := zoneOpCall.Do()
+		if err != nil {
+			errChan <- err
+			return multistep.ActionHalt
+		}
+
+		if newOp.Status == "RUNNING" {
```

Comment:
Is it possible that an operation for some reason goes from running to done without us noticing, so this branch never triggers? Maybe it should be `newOp.Status == "RUNNING" || newOp.Status == "DONE"`? I'm not super-familiar with the GCE API, so I may be misunderstanding things, though.